logging/env_logger.h

```diff
@@ -0,0 +1,177 @@
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include "port/sys_time.h"
+#include <time.h>
+
+#include "monitoring/iostats_context_imp.h"
+#include "rocksdb/env.h"
+#include "rocksdb/slice.h"
+#include "test_util/sync_point.h"
+#include "util/file_reader_writer.h"
+#include "util/mutexlock.h"
+
+namespace rocksdb {
+
+class EnvLogger : public Logger {
+ public:
+  EnvLogger(std::unique_ptr<WritableFileWriter>&& file, uint64_t(gettid)(),
+            Env* env, InfoLogLevel log_level = InfoLogLevel::ERROR_LEVEL)
+      : Logger(log_level),
+        file_(std::move(file)),
+        gettid_(gettid),
+        last_flush_micros_(0),
+        env_(env),
+        flush_pending_(false) {}
+
+  EnvLogger(const std::string& fname, const EnvOptions& options,
+            uint64_t(gettid)(), Env* env,
+            InfoLogLevel log_level = InfoLogLevel::ERROR_LEVEL)
+      : Logger(log_level),
+        gettid_(gettid),
+        last_flush_micros_(0),
+        env_(env),
+        flush_pending_(false) {
+    std::unique_ptr<WritableFile> writable_file;
+    env_->NewWritableFile(fname, &writable_file, options);
+    file_ = std::unique_ptr<WritableFileWriter>(
+        new WritableFileWriter(std::move(writable_file), fname, options, env));
+  }
+
+  ~EnvLogger() {
+    if (!closed_) {
+      closed_ = true;
+      CloseHelper();
+    }
+  }
+
+ private:
+  void FlushLocked() {
+    mutex_.AssertHeld();
+    if (flush_pending_) {
+      flush_pending_ = false;
+      file_->Flush();
+    }
+    last_flush_micros_ = env_->NowMicros();
+  }
+
+  void Flush() override {
+    TEST_SYNC_POINT("EnvLogger::Flush:Begin1");
+    TEST_SYNC_POINT("EnvLogger::Flush:Begin2");
+
+    MutexLock l(&mutex_);
+    FlushLocked();
+  }
+
+  Status CloseImpl() override { return CloseHelper(); }
+
+  Status CloseHelper() {
+    mutex_.Lock();
+    const auto close_status = file_->Close();
+    mutex_.Unlock();
+
+    if (close_status.ok()) {
+      return close_status;
+    }
+    return Status::IOError("Close of log file failed with error:" +
+                           (close_status.getState()
+                                ? std::string(close_status.getState())
+                                : std::string()));
+  }
+
+  using Logger::Logv;
+  void Logv(const char* format, va_list ap) override {
+    IOSTATS_TIMER_GUARD(logger_nanos);
+
+    const uint64_t thread_id = (*gettid_)();
+
+    // We try twice: the first time with a fixed-size stack allocated buffer,
+    // and the second time with a much larger dynamically allocated buffer.
+    char buffer[500];
+    for (int iter = 0; iter < 2; iter++) {
+      char* base;
+      int bufsize;
+      if (iter == 0) {
+        bufsize = sizeof(buffer);
+        base = buffer;
+      } else {
+        bufsize = 65536;
+        base = new char[bufsize];
+      }
+      char* p = base;
+      char* limit = base + bufsize;
+
+      struct timeval now_tv;
+      gettimeofday(&now_tv, nullptr);
```

Comment:
This might be Linux-only function, which makes the function not-portable. Can we use env_->NowMicros() instead? I know there might be challenges to print human readable log lines, but it's worth a try.

Reply:
Take back what I said. I looks like Windows supports it too. I saw win_logger.cc does the same, so it should be fine.